state/protocol/badger/mutator.go

```diff
@@ -835,6 +827,25 @@ func (m *FollowerState) epochMetricsAndEventsOnBlockFinalized(parentEpochState,
 	parentEpochPhase := parentEpochState.EpochPhase()
 	childEpochPhase := finalizedEpochState.EpochPhase()
 
+	// Check for entering or exiting EFM
+	if !parentEpochState.EpochFallbackTriggered() && finalizedEpochState.EpochFallbackTriggered() {
+		// this block triggers EFM
+		events = append(events, m.consumer.EpochFallbackModeTriggered)
+		metrics = append(metrics, m.metrics.EpochFallbackModeTriggered)
+	}
+	if parentEpochState.EpochFallbackTriggered() && !finalizedEpochState.EpochFallbackTriggered() {
+		// this block exits EFM
+		events = append(events, m.consumer.EpochFallbackModeExited)
+		metrics = append(metrics, m.metrics.EpochFallbackModeExited)
+	}
+
+	// Check for a new epoch extension
+	if len(finalizedEpochState.EpochExtensions()) > len(parentEpochState.EpochExtensions()) {
+		finalizedExtension := finalizedEpochState.EpochExtensions()[len(parentEpochState.EpochExtensions())]
```

Comment:
This code assumes that a single extension is added, but we don't check that in if statement meaning if multiple extensions were added(not the case currently but who knows) we will report only single event. I would propose or to report all extensions or to modify the check to ensure that single extension was added.

Reply:
Good point, updated in https://github.com/onflow/flow-go/pull/6062/commits/b03776b5cdae581a3adf021d49413642472f5413